### ISSUE TITLE
Support accessing Postgres Materialized Views as tables

### ISSUE THREAD

**Issue by TheDahv** [metabase/metabase#2355]
Is there anything special I need to know about Postgres' [materialized view](http://www.postgresql.org/docs/9.3/static/sql-creatematerializedview.html) feature?

I had a view set up that I was able to import when I ran the "Sync" command, but it is rather large and ended up being too slow. So I rebuilt everything to instead provide a materialized view with indexes and tried resyncing Postgres. Now the table is gone, and doesn't appear as an option after syncing.

I also made sure the table permissions were the same as the other tables that _do_ show up. I wasn't able to learn anything more by reading through the code to tell me whether this feature is supported or not, so I don't know further how to rule this out as my issue or as one in Metabase.

Here is some basic info about my setup:

* Metabase@v0.15.0
* Postgres@9.5.1
* Running in a Docker container
* Metabase and Postgres hosted on Ubuntu VM

Thanks again for the work so far!

**Comment by TheDahv** [comment]
I just saw https://github.com/metabase/metabase/issues/2329. I don't suspect that is my problem as I made the schema changes on Friday 8 April. However, if there is something I can run locally to determine if I'm still syncing, that would be helpful to know.

**Comment by agilliland** [comment]
Hmmm, I don't recall every specifically testing against materialized views, but I see no reason why it wouldn't be working.

If you can provide me with a very simple working example of how you are building your materialized view I can try to replicate that locally and make sure it's working for me.

Also, if you check out your Metabase logs there might be some clues in there.  On a docker container you should be able to just run `docker logs <container-name|container-id>` and get the output.

**Comment by TheDahv** [comment]
Here's a simple view with some fake data:

```
DROP MATERIALIZED VIEW IF EXISTS test_view;
CREATE MATERIALIZED VIEW test_view AS
SELECT 1 AS id, 'David' AS name, 'TheDahv' AS handle, to_date('2009-04-13', 'YYYY-MM-DD') AS user_since
UNION ALL
SELECT 2, 'Allen', 'agilliland', to_date('2012-07-16', 'YYYY-MM-DD');
CREATE INDEX user_since ON test_view (user_since);
```

The real example is a large `SELECT` query joining across multiple tables, but I _think_ this should convey the basic idea.

**Comment by TheDahv** [comment]
Here are some logs from the last sync. It doesn't even mention the view I'm having trouble with (`duplicates_progress`):

```
04-12 06:00:06 DEBUG models.interface :: DB CALL: Database 2
04-12 06:00:06 DEBUG db.internal :: DB CALL: User [:id :email :date_joined :first_name :last_name :last_login :is_superuser :is_qbnewb :is_active :is_staff] (where (limit (select* ENTITY24489) 1) {:id current-user-id})
04-12 06:00:06 DEBUG metabase.middleware :: POST /api/database/2/sync 200 (5 ms)
04-12 06:00:06 DEBUG db.internal :: DB CALL: Database * (where (limit (select* ENTITY29138) 1) {:id (events/object->model-id topic object)})
04-12 06:00:06 INFO driver.sync :: Syncing PostgreSQL database 'Moz Local ETL DB'...
04-12 06:00:06 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:07 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:07 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:07 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:08 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:08 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:08 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:09 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:09 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:09 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG driver.sync :: [****??????????????????????????????????????????????] ?    8% Synced table 'duplicates'.
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG driver.sync :: [********??????????????????????????????????????????] ?   17% Synced table 'duplicates_raw'.
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG driver.sync :: [************??????????????????????????????????????] ?   25% Synced table 'enterprise_locations'.
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:10 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG driver.sync :: [****************??????????????????????????????????] ?   33% Synced table 'enterprise_users'.
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG driver.sync :: [********************??????????????????????????????] ?   42% Synced table 'insights_updates'.
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG driver.sync :: [*************************?????????????????????????] ?   50% Synced table 'locations'.
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:11 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG driver.sync :: [*****************************?????????????????????] ?   58% Synced table 'locations_accuracy'.
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG driver.sync :: [*********************************?????????????????] ?   67% Synced table 'locations_raw'.
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG driver.sync :: [*************************************?????????????] ?   75% Synced table 'managed_locations'.
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:12 DEBUG driver.sync :: [*****************************************?????????] ?   83% Synced table 'partners'.
04-12 06:00:12 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG driver.sync :: [*********************************************?????] ?   92% Synced table 'users'.
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG generic-sql.query-processor :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:00:13 DEBUG driver.sync :: [**************************************************] ?  100% Synced table 'users_properties'.
04-12 06:00:13 INFO driver.sync :: Finished syncing PostgreSQL database 'Moz Local ETL DB'. (7 s)
```

**Comment by TheDahv** [comment]
Here's the last clue: I just tried writing a raw SQL query by hand, and it worked.

Here are the logs when the query runs:

```
04-12 06:07:29 DEBUG driver.query-processor ::
QUERY: ?
{:database 2, :type "native", :native {:query "SELECT *\nFROM duplicates_progress\nLIMIT 20"}, :constraints {:max-results 10000, :max-results-bare-rows 2000}}

04-12 06:07:29 DEBUG db.internal :: DB CALL: Database [:engine :details] (where (limit (select* ENTITY27565) 1) {:id database-id})
04-12 06:07:29 DEBUG generic-sql.native :: UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';
04-12 06:07:29 DEBUG generic-sql.native :: SELECT *
FROM duplicates_progress
LIMIT 20
04-12 06:07:29 DEBUG models.interface :: DB CALL: QueryExecution 3474
04-12 06:07:29 DEBUG metabase.middleware :: POST /api/dataset 200 (23 ms)
04-12 06:07:29 WARN metabase.driver :: Don't know how to map class 'null' to a Field base_type, falling back to :UnknownField.
04-12 06:07:29 WARN metabase.driver :: Don't know how to map class 'null' to a Field base_type, falling back to :UnknownField.
```

**Comment by agilliland** [comment]
Awesome.  Thanks for all the info.  I'll try to replicate this today and see what I find.

Based on what I'm seeing it looks like our sync process is somehow not seeing materialized view tables because as you said, it's not even being listed in the sync process.

The SQL writing interface just provides an open connection to the database and has no restrictions or real insight into what's in the database, so it would be able to access anything you can access directly from the psql console.  Makes sense that you'd be able to access your table that way and validates that it's really there ;)

**Comment by TheDahv** [comment]
Right. That SQL was based on queries I wrote against the view _before_ I upgraded it to a materialized view.

FWIW, I haven't worked with `psql` in a few years and didn't even know materialized views existed until I asked myself "how can I make this view go faster!?". It turns out that they're crazy fast, but just not too accessible to my colleagues outside of the engineering team.

Again, thank you for the work so far and for your responsiveness. We're loving it so far :)

**Comment by agilliland** [comment]
Quick follow up here, I verified that the database introspection process is indeed not seeing the postgres materialized views which is why your table isn't showing up.  So we'll need to make some updates to that process so that we find and identify those tables.

**Comment by TheDahv** [comment]
👍 thanks for looking into it

**Comment by agilliland** [comment]
Quick update here.  I looked into this and unfortunately the Postgres team made some perhaps strange choices and purposely kept materialized views out of any schema information lookups that are normally used ... http://www.postgresql.org/message-id/flat/20141011004406.GQ28859@tamriel.snowman.net#20141011004406.GQ28859@tamriel.snowman.net

It looks like in order to include these we'd have to go directly to the `pg_matviews` system catalog, which is pretty specific.

**Comment by rafaelveloso** [comment]
> It looks like in order to include these we'd have to go directly to the pg_matviews system catalog, which is pretty specific.

Still, materialized views are something are very used to speedup queries that require a lot of data and computation (typical aggregations and lookups for analytics), so IMO this is something that might be worth to implement even if it's too specific.

**Comment by TheDahv** [comment]
Yeah, I didn't interpret that as "we're not going to to do it"--especially since the issue is still open.

You're right, though, that this does make querying over projected data very quick.

**Comment by camsaul** [comment]
We already dig right into the postgres system catalogs to get foreign key info for Redshift since it doesn't work with JDBC schema methods either. So we could follow a similar pattern to get materialized views

**Comment by camsaul** [comment]
Implemented by #2559

**Comment by TheDahv** [comment]
woohoo! Thanks @camsaul 

Can't wait to give it a try.
